File: scripts/day9-part2-simple.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <chrono>
#include <sstream>

using namespace std;

struct Point {
    int x, y;
};

struct Edge {
    int minX, minY, maxX, maxY;
};

vector<Edge> getNormalizedEdges(const vector<Point>& tiles) {
    vector<Edge> edges;
    int n = tiles.size();
    
    for (int i = 0; i < n - 1; i++) {
        edges.push_back({
            min(tiles[i].x, tiles[i+1].x),
            min(tiles[i].y, tiles[i+1].y),
            max(tiles[i].x, tiles[i+1].x),
            max(tiles[i].y, tiles[i+1].y)
        });
    }
    
    // Wrap around from last to first
    edges.push_back({
        min(tiles[n-1].x, tiles[0].x),
        min(tiles[n-1].y, tiles[0].y),
        max(tiles[n-1].x, tiles[0].x),
        max(tiles[n-1].y, tiles[0].y)
    });
    
    return edges;
}

bool isFullyContained(const vector<Edge>& edges, int minX, int minY, int maxX, int maxY) {
    for (const auto& edge : edges) {
        if (minX < edge.maxX && maxX > edge.minX && minY < edge.maxY && maxY > edge.minY) {
            return false;
        }
    }
    return true;
}

long long solvePart2(const vector<Point>& tiles) {
    int n = tiles.size();
    long long maxArea = 0;
    vector<Edge> edges = getNormalizedEdges(tiles);
    
    for (int i = 0; i < n; i++) {
        for (int j = i + 1; j < n; j++) {
            long long area = ((long long)abs(tiles[i].x - tiles[j].x) + 1) * 
                            (abs(tiles[i].y - tiles[j].y) + 1);
            
            if (area <= maxArea) continue;
            
            int minX = min(tiles[i].x, tiles[j].x);
            int maxX = max(tiles[i].x, tiles[j].x);
            int minY = min(tiles[i].y, tiles[j].y);
            int maxY = max(tiles[i].y, tiles[j].y);
            
            if (isFullyContained(edges, minX, minY, maxX, maxY)) {
                maxArea = area;
            }
        }
    }
    
    return maxArea;
}

int main() {
    ifstream file("../inputs/day-9-input.txt");
    vector<Point> tiles;
    string line;
    
    while (getline(file, line)) {
        stringstream ss(line);
        string token;
        Point p;
        
        getline(ss, token, ',');
        p.x = stoi(token);
        getline(ss, token, ',');
        p.y = stoi(token);
        
        tiles.push_back(p);
    }
    file.close();
    
    const int iterations = 1;
    auto start = chrono::high_resolution_clock::now();
    
    long long result;
    for (int i = 0; i < iterations; i++) {
        result = solvePart2(tiles);
    }
    
    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::nanoseconds>(end - start);
    double avgMicros = duration.count() / (double)iterations / 1000.0;
    
    cout << "C++ (Day 9 Part 2 - Simple)" << endl;
    cout << "Answer: " << result << endl;
    cout << "Average time: " << avgMicros << " microseconds (" << iterations << " iterations)" << endl;
    cout << "Total time: " << (duration.count() / 1000000.0) << " ms" << endl;
    
    return 0;
}

File: scripts/day9-part2-simple.js
const fs = require("fs");

function getNormalizedEdges(tiles) {
  const edges = [];
  const n = tiles.length;

  for (let i = 0; i < n - 1; i++) {
    edges.push({
      minX: Math.min(tiles[i][0], tiles[i + 1][0]),
      minY: Math.min(tiles[i][1], tiles[i + 1][1]),
      maxX: Math.max(tiles[i][0], tiles[i + 1][0]),
      maxY: Math.max(tiles[i][1], tiles[i + 1][1]),
    });
  }

  // Wrap around from last to first
  edges.push({
    minX: Math.min(tiles[n - 1][0], tiles[0][0]),
    minY: Math.min(tiles[n - 1][1], tiles[0][1]),
    maxX: Math.max(tiles[n - 1][0], tiles[0][0]),
    maxY: Math.max(tiles[n - 1][1], tiles[0][1]),
  });

  return edges;
}

function isFullyContained(edges, minX, minY, maxX, maxY) {
  for (const edge of edges) {
    if (
      minX < edge.maxX &&
      maxX > edge.minX &&
      minY < edge.maxY &&
      maxY > edge.minY
    ) {
      return false;
    }
  }
  return true;
}

function solvePart2(tiles) {
  const n = tiles.length;
  let maxArea = 0;
  const edges = getNormalizedEdges(tiles);

  for (let i = 0; i < n; i++) {
    for (let j = i + 1; j < n; j++) {
      const area =
        (Math.abs(tiles[i][0] - tiles[j][0]) + 1) *
        (Math.abs(tiles[i][1] - tiles[j][1]) + 1);

      if (area <= maxArea) continue;

      const minX = Math.min(tiles[i][0], tiles[j][0]);
      const maxX = Math.max(tiles[i][0], tiles[j][0]);
      const minY = Math.min(tiles[i][1], tiles[j][1]);
      const maxY = Math.max(tiles[i][1], tiles[j][1]);

      if (isFullyContained(edges, minX, minY, maxX, maxY)) {
        maxArea = area;
      }
    }
  }

  return maxArea;
}

const input = fs
  .readFileSync("../inputs/day-9-input.txt", "utf-8")
  .trim()
  .split(/\r?\n/);

const tiles = input.map((line) => line.split(",").map(Number));

const iterations = 1;
const startTime = process.hrtime.bigint();

let result;
for (let i = 0; i < iterations; i++) {
  result = solvePart2(tiles);
}

const endTime = process.hrtime.bigint();
const totalNanos = Number(endTime - startTime);
const avgMicros = (totalNanos / iterations / 1000).toFixed(3);

console.log(`JavaScript (Day 9 Part 2 - Simple)`);
console.log(`Answer: ${result}`);
console.log(
  `Average time: ${avgMicros} microseconds (${iterations} iterations)`
);
console.log(`Total time: ${(totalNanos / 1000000).toFixed(2)} ms`);
